src/Features/Core/Portable/ConvertAutoPropertyToFullProperty/AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+
+namespace Microsoft.CodeAnalysis.ConvertAutoPropertyToFullProperty
+{
+    internal abstract class AbstractConvertAutoPropertyToFullPropertyCodeRefactoringProvider
+        : CodeRefactoringProvider
+    {
+        internal abstract SyntaxNode GetProperty(SyntaxToken token);
+        internal abstract Task<string> GetFieldNameAsync(Document document, IPropertySymbol propertySymbol, CancellationToken cancellationToken);
+        internal abstract (SyntaxNode newGetAccessor, SyntaxNode newSetAccessor) GetNewAccessors(
+            DocumentOptionSet options, SyntaxNode property, string fieldName, SyntaxGenerator generator);
+        internal abstract SyntaxNode GetPropertyWithoutInitializer(SyntaxNode property);
+        internal abstract SyntaxNode GetInitializerValue(SyntaxNode property);
+        internal abstract SyntaxNode ConvertPropertyToExpressionBodyIfDesired(DocumentOptionSet options, SyntaxNode fullProperty);
+        internal abstract SyntaxNode GetTypeBlock(SyntaxNode syntaxNode);
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            var property = GetProperty(token);
+            if (property == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var propertySymbol = semanticModel.GetDeclaredSymbol(property) as IPropertySymbol;
+            if (propertySymbol == null)
+            {
+                return;
+            }
+
+            if (!(IsValidAutoProperty(property, propertySymbol)))
+            {
+                return;
+            }
+
+            context.RegisterRefactoring(
+                new ConvertAutoPropertyToFullPropertyCodeAction(
+                    FeaturesResources.Convert_to_full_property,
```

Comment:
title doesn't need to be supplied here.  it can just be inlined in the constructor for ConvertAutoPropertyToFullPropertyCodeAction